src/functions.php

```diff
@@ -341,43 +341,67 @@ function _checkTypehint(callable $callback, $object)
         return true;
     }
 
-    if (\PHP_VERSION_ID < 70100 || \defined('HHVM_VERSION')) {
-        $expectedException = $parameters[0];
+    $expectedException = $parameters[0];
 
+    // PHP before v8 used an easy API:
+    if (\PHP_VERSION_ID < 70100 || \defined('HHVM_VERSION')) {
         if (!$expectedException->getClass()) {
             return true;
         }
 
         return $expectedException->getClass()->isInstance($object);
-    } else {
-        $type = $parameters[0]->getType();
+    }
 
-        if (!$type) {
-            return true;
-        }
+    // Extract the type of the argument and handle different possibilities
+    $type = $expectedException->getType();
+    
+    $isTypeUnion = true;
+    $types = [];
+
+    switch (true) {
+        case $type === null:
+            break;
+        case $type instanceof \ReflectionNamedType:
+            $types = [$type];
+            break;
+        case $type instanceof \ReflectionIntersectionType:
+            $isTypeUnion = false;
```

Comment:
Missing break here?

Reply:
No, this is intentional. Intersection and union both operate on the type-list just with and Vs or. We can add a comment though that says so, or be explicit about the code. I don't have a strong opinion about it. 🤷‍♂️